### DISCUSSION TITLE
<img> tag onError function doesn't work

### DISCUSSION THREAD

**Discussion by aliemami** [vercel/next.js#14772] [Help]
# Bug report

## Describe the bug
I'm trying to render an image if it exist and not show it if it doesn't, same code in reactjs using the browser side rendering works but it fails to work with using nextjs. here is a link to the sandbox 

nextjs example in the sandbox 
https://codesandbox.io/s/nextjs-3jlqh?file=/pages/index.js

server side rendering sandbox 
https://codesandbox.io/s/elated-kapitsa-h8zfi?file=/src/App.js

## To Reproduce
from nodejs 

         app.get('/test', (request, response) => {
             return next.render(request, response, '/image/testImage', {})
         });

nextjs component 

      import React from "react";

    export default function TestImage() {
      const replaceImgWithError = e => {
         e.target.onerror = null;
         e.target.src = "http://placekitten.com/g/200/300";
     };

     const hideImgWhenError = e => {
         e.target.onerror = null;
         e.target.style.display = "none";
     };

    return (
        <div>
            <h1>Hello CodeSandbox</h1>
            <h2>Start editing to see some magic happen!</h2>
            <img
                onError={replaceImgWithError}
                alt="foo"
                src="https://crablets-event-images.s3.amazonaws.com/users/10/top"
            />
            <img
                onError={hideImgWhenError}
                alt="foo"
                src="https://crablets-event-images.s3.amazonaws.com/users/10/top"
            />
        </div>
    );
}

Error 

it shows broken images 

## Expected behavior

the onError functions should be invoked and replace the image with default or change the css so the image doesn't show 

## Screenshots

USING NEXTJS 
<img width="590" alt="Screen Shot 2020-07-01 at 10 25 08 AM" src="https://user-images.githubusercontent.com/382247/86255519-3441e680-bb85-11ea-8252-d1784e8ceaa4.png">

WITH BROWSER SIDE RENDERING 
<img width="528" alt="Screen Shot 2020-07-01 at 10 25 48 AM" src="https://user-images.githubusercontent.com/382247/86255994-cba73980-bb85-11ea-94eb-1551e8ef86e3.png">

## System information

- OS: persistant through different OS's including mac and ubuntu 
- Version of Next.js: v9.4.5-canary.26
- Version of Node.js:  v14.0.0

**Comment by jamesmosier** [comment]
As you mentioned, this will not work with server side rendering. 

I created a new file:

```jsx
export default function Images() {
  const replaceImgWithError = e => {
    e.target.onerror = null;
    e.target.src = 'http://placekitten.com/g/200/300';
  };

  const hideImgWhenError = e => {
    e.target.onerror = null;
    e.target.style.display = 'none';
  };

  return (
    <>
      <p>hi</p>
      <img
        onError={replaceImgWithError}
        alt="foo"
        src="https://crablets-event-images.s3.amazonaws.com/users/10/top"
      />
      <img
        onError={hideImgWhenError}
        alt="foo"
        src="https://crablets-event-images.s3.amazonaws.com/users/10/top"
      />
    </>
  );
}

```

and then imported it in `pages/index.js` like:

```jsx
const Images = dynamic(() => import('../src/Images'), { ssr: false });
export default function IndexPage() {
  return (
    <div>
      <h1>Hello CodeSandbox</h1>
      <h2>Start editing to see some magic happen!</h2>
      <Images />
    </div>
  );
}
```

and it worked using [next/dynamic](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr)

[codesandbox example](https://codesandbox.io/s/nextjs-gvubb)

**Comment by kwaoquaye** [comment]
After little debugging, I realized onError does not trigger when you use require to import your images within your component. So to handle it without throwing an error and ending process wrap the import(i.e using require) in a try catch and return a default image should it fail. Below is an illustration:

```
import defaultImg from "assets/icons/default.png";

const safeImgImport = (itemSize) => {
    try {
      let temp = require(`../../../.././assets/icons/${itemSize}.png`).default
      return temp
    } catch (error) {
      return defaultImg
    }
  }

const Image = (Item)=> {
     return (
<img
          src={safeImgImport()}
          alt="Nice Img"
        />
     )
}
```

And it worked too in Reactjs

**Comment by TanyaZhecheva** [comment]
Im using Image from "next/image";

i used to have 403 error when passing url to src
to fix this i just added 'unoptimized'

example:
<Image
src={'https://hudabeauty.com/cdn/shop/files/PDP-SECTION1-FFCONCEALER-WHIPPEDCREAM-TILE3.webp?  v=1759869562&width=1946'}
fill
unoptimized
priority
/>

hope this helps

**Comment by Niteshagarwal01** [comment]
***

Hey there! 👋 This is a classic issue that pops up when mixing Server-Side Rendering (SSR) with React's event system. 

### Why this is happening (The SSR Race Condition)
The reason this works perfectly in purely client-side React but fails in Next.js boils down to a **race condition** during the hydration phase.

1. **In Next.js (SSR):** The server sends down the initial HTML with the `<img src="...">` tag. As soon as the browser parses this HTML, it immediately tries to download the image. If the image link is broken, the browser fires a native `error` event almost instantly. Meanwhile, React is still busy downloading the JS bundle and booting up (hydrating). If the native `error` event fires *before* React finishes hydrating, **React misses the event entirely** and your `onError` handler is never called.
2. **In Standard React (CSR):** The `<img>` tag doesn't exist until React creates it. When React inserts the image into the DOM, it attaches the `onError` listener at the exact same millisecond. Therefore, it's impossible for React to miss the event.

### The Fix
Since we can't guarantee React will hydrate before the image fails, we need to manually check if the image has already errored by the time the component mounts. 

We can do this safely by checking the `complete` and `naturalWidth` properties of the image via a `ref`. If an image is `complete` but its `naturalWidth` is `0`, we know it's a broken image.

Here is a robust, SSR-safe `ImageWithFallback` component you can use to solve this:

```jsx
import React, { useState, useEffect, useRef } from "react";

// Usage Example
export default function TestImage() {
  return (
    <div>
      {/* Replaces image with a default on error */}
      <ImageWithFallback
        src="https://crablets-event-images.s3.amazonaws.com/users/10/top"
        fallbackSrc="http://placekitten.com/g/200/300"
        alt="User Avatar"
      />

      {/* Hides the image entirely on error */}
      <ImageWithFallback
        src="https://crablets-event-images.s3.amazonaws.com/users/10/top"
        alt="Hidden Image"
        hideOnError={true}
      />
    </div>
  );
}

// The Component
function ImageWithFallback({ src, fallbackSrc, alt, hideOnError = false }) {
  const [hasErrored, setHasErrored] = useState(false);
  const imgRef = useRef(null);

  useEffect(() => {
    // When the component mounts, check if the image already 
    // failed to load before React hydrated.
    if (imgRef.current) {
      const img = imgRef.current;
      if (img.complete && img.naturalWidth === 0) {
        setHasErrored(true);
      }
    }
  }, [src]);

  if (hasErrored) {
    if (hideOnError) return null; // Hide the image completely
    return <img src={fallbackSrc} alt={alt} />; // Show fallback
  }

  return (
    <img
      ref={imgRef}
      src={src}
      alt={alt}
      onError={() => setHasErrored(true)}
    />
  );
}
```

### Note on Modern Next.js
I noticed from your system info you are on `v9.4.5`. If you ever upgrade to Next.js 10+, I highly recommend switching to the built-in `next/image` component (`import Image from 'next/image'`). Vercel built it specifically to handle these SSR edge cases natively under the hood!

Hope this helps solve the mystery for you! Let me know if you have any other questions. 🚀

***

**Comment by abduromanov2020** [comment]
ran into this too. the issue is that the browser loads (and fails) images before React hydrates, so by the time `onError` is attached, the error already happened and won't fire again.

what worked for me is checking the image state after mount:

```tsx
function ImageWithFallback({ src, fallback, alt, ...props }) {
  const imgRef = useRef(null);
  const [imgSrc, setImgSrc] = useState(src);

  useEffect(() => {
    const img = imgRef.current;
    if (img?.complete && img.naturalWidth === 0) {
      setImgSrc(fallback);
    }
  }, [fallback]);

  return (
    <img
      ref={imgRef}
      src={imgSrc}
      alt={alt}
      onError={() => setImgSrc(fallback)}
      {...props}
    />
  );
}
```

the `useEffect` catches images that already failed before hydration (`complete` is true but `naturalWidth` is 0), and `onError` handles anything that fails after. covers both cases.

`next/image` also handles this out of the box if you can use it.
